Core/DISET/private/Service.py

```diff
@@ -317,264 +385,298 @@ def _processInThread( self, clientTransport ):
+    # Notify the client we're ready to execute the action
+    retVal = self._transportPool.send(trid, S_OK())
+    if not retVal['OK']:
       return retVal
 
     messageConnection = False
-    if proposalTuple[1] == ( 'Connection', 'new' ):
+    if proposalTuple[1] == ('Connection', 'new'):
       messageConnection = True
 
     if messageConnection:
 
       if self._msgBroker.getNumConnections() > self._cfg.getMaxMessagingConnections():
-        result = S_ERROR( "Maximum number of connections reached. Try later" )
-        result[ 'closeTransport' ] = True
+        result = S_ERROR("Maximum number of connections reached. Try later")
+        result['closeTransport'] = True
         return result
 
-      #This is a stable connection
-      self._msgBroker.addTransportId( trid, self._name,
-                                      receiveMessageCallback = self._mbReceivedMsg,
-                                      disconnectCallback = self._mbDisconnect,
-                                      listenToConnection = False )
+      # This is a stable connection
+      self._msgBroker.addTransportId(trid, self._name,
+                                     receiveMessageCallback=self._mbReceivedMsg,
+                                     disconnectCallback=self._mbDisconnect,
+                                     listenToConnection=False)
 
-    result = self._executeAction( trid, proposalTuple, handlerObj )
-    if result[ 'OK' ] and messageConnection:
-      self._msgBroker.listenToTransport( trid )
-      result = self._mbConnect( trid, handlerObj )
-      if not result[ 'OK' ]:
-        self._msgBroker.removeTransport( trid )
+    result = self._executeAction(trid, proposalTuple, handlerObj)
+    if result['OK'] and messageConnection:
+      self._msgBroker.listenToTransport(trid)
+      result = self._mbConnect(trid, handlerObj)
+      if not result['OK']:
+        self._msgBroker.removeTransport(trid)
 
-    result[ 'closeTransport' ] = not messageConnection or not result[ 'OK' ]
+    result['closeTransport'] = not messageConnection or not result['OK']
     return result
 
-  def _mbConnect( self, trid, handlerObj = None ):
+  def _mbConnect(self, trid, handlerObj=None):
     if not handlerObj:
-      result = self._instantiateHandler( trid )
-      if not result[ 'OK' ]:
+      result = self._instantiateHandler(trid)
+      if not result['OK']:
         return result
-      handlerObj = result[ 'Value' ]
-    return handlerObj._rh_executeConnectionCallback( 'connected' )
+      handlerObj = result['Value']
+    return handlerObj._rh_executeConnectionCallback('connected')
 
-  def _executeAction( self, trid, proposalTuple, handlerObj ):
+  def _executeAction(self, trid, proposalTuple, handlerObj):
     try:
-      return handlerObj._rh_executeAction( proposalTuple )
+      response = handlerObj._rh_executeAction(proposalTuple)
+      if self.activityMonitoring and response["OK"]:
+        self.activityMonitoringReporter.addRecord({
```

Comment:
Here, you are back in a thread, which we already concluded is not really thread safe. 
Not to waste more time, I'll ask @zmathe to make it thread safe (it should be quite straight forward)